sdk/src/phoenix/phoenixSubscriber.ts

```diff
@@ -0,0 +1,154 @@
+import { Connection, PublicKey, SYSVAR_CLOCK_PUBKEY } from '@solana/web3.js';
+import { BulkAccountLoader } from '../accounts/bulkAccountLoader';
+import {
+	MarketData,
+	Client,
+	deserializeMarketData,
+	deserializeClockData,
+	toNum,
+	getMarketUiLadder,
+} from '@ellipsis-labs/phoenix-sdk';
+
+export type PhoenixMarketSubscriberConfig = {
+	connection: Connection;
+	programId: PublicKey;
+	marketAddress: PublicKey;
+	accountSubscription:
+		| {
+				// enables use to add web sockets in the future
+				type: 'polling';
+				accountLoader: BulkAccountLoader;
+		  }
+		| {
+				type: 'websocket';
+		  };
+};
+
+export class PhoenixSubscriber {
+	connection: Connection;
+	client: Client;
+	programId: PublicKey;
+	marketAddress: PublicKey;
+	subscriptionType: 'polling' | 'websocket';
+	accountLoader: BulkAccountLoader | undefined;
+	market: MarketData;
+	marketCallbackId: string | number;
+	clockCallbackId: string | number;
+
+	subscribed: boolean;
+	lastSlot: number;
+	lastUnixTimestamp: number;
+
+	public constructor(config: PhoenixMarketSubscriberConfig) {
+		this.connection = config.connection;
+		this.programId = config.programId;
+		this.marketAddress = config.marketAddress;
+		if (config.accountSubscription.type === 'polling') {
+			this.subscriptionType = 'polling';
+			this.accountLoader = config.accountSubscription.accountLoader;
+		} else {
+			this.subscriptionType = 'websocket';
+		}
+		this.lastSlot = 0;
+		this.lastUnixTimestamp = 0;
+	}
+
+	public async subscribe(): Promise<void> {
+		if (this.subscribed) {
+			return;
+		}
+
+		this.market = deserializeMarketData(
+			(await this.connection.getAccountInfo(this.marketAddress, 'confirmed'))
+				.data
+		);
+
+		const clock = deserializeClockData(
+			(await this.connection.getAccountInfo(SYSVAR_CLOCK_PUBKEY, 'confirmed'))
+				.data
+		);
+		this.lastUnixTimestamp = toNum(clock.unixTimestamp);
+
+		if (this.subscriptionType === 'websocket') {
+			this.marketCallbackId = this.connection.onAccountChange(
+				this.marketAddress,
+				(accountInfo, _ctx) => {
+					this.market = deserializeMarketData(accountInfo.data);
+				}
+			);
+			this.clockCallbackId = this.connection.onAccountChange(
+				SYSVAR_CLOCK_PUBKEY,
+				(accountInfo, ctx) => {
+					this.lastSlot = ctx.slot;
+					const clock = deserializeClockData(accountInfo.data);
+					this.lastUnixTimestamp = toNum(clock.unixTimestamp);
+				}
+			);
+		} else {
+			this.marketCallbackId = await this.accountLoader.addAccount(
+				this.marketAddress,
+				(buffer, slot) => {
+					this.marketCallbackId = slot;
+					this.market = deserializeMarketData(buffer);
+				}
+			);
+			this.clockCallbackId = await this.accountLoader.addAccount(
+				SYSVAR_CLOCK_PUBKEY,
+				(buffer, slot) => {
+					this.clockCallbackId = slot;
+					const clock = deserializeClockData(buffer);
+					this.lastUnixTimestamp = toNum(clock.unixTimestamp);
+				}
+			);
+		}
+
+		this.subscribed = true;
+	}
+
+	public getBestBid(): number {
```

Comment:
I'm having this return the price in human readable units. Lmk if you prefer something different

Reply:
this and `getBestAsk` should be in `PRICE_PRECISION` to be consistent with `SerumSubscriber`

https://github.com/drift-labs/protocol-v2/blob/006b2694c67c06df4f97a2ef4e607b7c531f5149/sdk/src/serum/serumSubscriber.ts#L112